File: jscomp/test/js_float_test.ml
open Js_float

let suites = Mt.[
  ("_NaN <> _NaN", (fun _ ->
    Eq(false, _NaN = _NaN)));

  ("isNaN - _NaN", (fun _ ->
    Eq(true, isNaN _NaN)));
  ("isNaN - 0.", (fun _ ->
    Eq(false, isNaN 0.)));

  ("isFinite - infinity", (fun _ ->
    Eq(false, isFinite infinity)));
  ("isFinite - neg_infinity", (fun _ ->
    Eq(false, isFinite neg_infinity)));
  ("isFinite - _NaN", (fun _ ->
    Eq(false, isFinite _NaN)));
  ("isFinite - 0.", (fun _ ->
    Eq(true, isFinite 0.)));

  ("toExponential", (fun _ ->
    Eq("1.23456e+2", toExponential 123.456)));
  ("toExponential - large number", (fun _ ->
    Eq("1.2e+21", toExponential 1.2e21)));
  ("toExponentialWithPrecision - digits:2", (fun _ ->
    Eq("1.23e+2", toExponentialWithPrecision 123.456 ~digits:2)));
  ("toExponentialWithPrecision - digits:4", (fun _ ->
    Eq("1.2346e+2", toExponentialWithPrecision 123.456 ~digits:4)));
  ("toExponentialWithPrecision - digits:20", (fun _ ->
    Eq("0.00000000000000000000e+0", toExponentialWithPrecision 0. ~digits:20)));
  (__LOC__, (fun _ ->
    ThrowAny(fun () -> ignore @@ toExponentialWithPrecision 0. ~digits:101)));
  ("toExponentialWithPrecision - digits:-1", (fun _ ->
    ThrowAny(fun () -> ignore @@ toExponentialWithPrecision 0. ~digits:(-1))));

  ("toFixed", (fun _ ->
    Eq("123", toFixed 123.456)));
  ("toFixed - large number", (fun _ ->
    Eq("1.2e+21", toFixed 1.2e21)));
  ("toFixedWithPrecision - digits:2", (fun _ ->
    Eq("123.46", toFixedWithPrecision 123.456 ~digits:2)));
  ("toFixedWithPrecision - digits:4", (fun _ ->
    Eq("123.4560", toFixedWithPrecision 123.456 ~digits:4)));
  ("toFixedWithPrecision - digits:20", (fun _ ->
    Eq("0.00000000000000000000", toFixedWithPrecision 0. ~digits:20)));
  ("toFixedWithPrecision - digits:101", (fun _ ->
    ThrowAny(fun () -> ignore @@ toFixedWithPrecision 0. ~digits:101)));
  ("toFixedWithPrecision - digits:-1", (fun _ ->
    ThrowAny(fun () -> ignore @@ toFixedWithPrecision 0. ~digits:(-1))));

  ("toPrecision", (fun _ ->
    Eq("123.456", toPrecision 123.456)));
  ("toPrecision - large number", (fun _ ->
    Eq("1.2e+21", toPrecision 1.2e21)));
  ("toPrecisionWithPrecision - digits:2", (fun _ ->
    Eq("1.2e+2", toPrecisionWithPrecision 123.456 ~digits:2)));
  ("toPrecisionWithPrecision - digits:4", (fun _ ->
    Eq("123.5", toPrecisionWithPrecision 123.456 ~digits:4)));
  ("toPrecisionWithPrecision - digits:20", (fun _ ->
    Eq("0.0000000000000000000", toPrecisionWithPrecision 0. ~digits:20)));
  (__LOC__, (fun _ ->
    ThrowAny(fun () -> ignore @@ toPrecisionWithPrecision 0. ~digits:101)));
  ("toPrecisionWithPrecision - digits:-1", (fun _ ->
    ThrowAny(fun () -> ignore @@ toPrecisionWithPrecision 0. ~digits:(-1))));

  ("toString", (fun _ ->
    Eq("1.23", toString 1.23)));
  ("toString - large number", (fun _ ->
    Eq("1.2e+21", toString 1.2e21)));
  ("toStringWithRadix - radix:2", (fun _ ->
    Eq( "1111011.0111010010111100011010100111111011111001110111",
        toStringWithRadix 123.456 ~radix:2)));
  ("toStringWithRadix - radix:16", (fun _ ->
    Eq("7b.74bc6a7ef9dc", toStringWithRadix 123.456 ~radix:16)));
  ("toStringWithRadix - radix:36", (fun _ ->
    Eq("3f", toStringWithRadix 123. ~radix:36)));
  ("toStringWithRadix - radix:37", (fun _ ->
    ThrowAny(fun () -> ignore @@ toStringWithRadix 0. ~radix:37)));
  ("toStringWithRadix - radix:1", (fun _ ->
    ThrowAny(fun () -> ignore @@ toStringWithRadix 0. ~radix:1)));
  ("toStringWithRadix - radix:-1", (fun _ ->
    ThrowAny(fun () -> ignore @@ toStringWithRadix 0. ~radix:(-1))));

  ("fromString - 123", (fun _ ->
    Eq(123., fromString "123")));
  ("fromString - 12.3", (fun _ ->
    Eq(12.3, fromString "12.3")));
  ("fromString - empty string", (fun _ ->
    Eq(0., fromString "")));
  ("fromString - 0x11", (fun _ ->
    Eq(17., fromString "0x11")));
  ("fromString - 0b11", (fun _ ->
    Eq(3., fromString "0b11")));
  ("fromString - 0o11", (fun _ ->
    Eq(9., fromString "0o11")));
  ("fromString - invalid string", (fun _ ->
    Eq(true, fromString "foo" |> isNaN)));
]

;; Mt.from_pair_suites __FILE__ suites

File: jscomp/test/js_float_test.js
'use strict';

var Mt = require("./mt.js");
var Pervasives = require("../../lib/js/pervasives.js");

var suites_0 = [
  "_NaN <> _NaN",
  (function (param) {
      return {
              TAG: /* Eq */0,
              _0: false,
              _1: NaN === NaN
            };
    })
];

var suites_1 = {
  hd: [
    "isNaN - _NaN",
    (function (param) {
        return {
                TAG: /* Eq */0,
                _0: true,
                _1: isNaN(NaN)
              };
      })
  ],
  tl: {
    hd: [
      "isNaN - 0.",
      (function (param) {
          return {
                  TAG: /* Eq */0,
                  _0: false,
                  _1: isNaN(0)
                };
        })
    ],
    tl: {
      hd: [
        "isFinite - infinity",
        (function (param) {
            return {
                    TAG: /* Eq */0,
                    _0: false,
                    _1: isFinite(Pervasives.infinity)
                  };
          })
      ],
      tl: {
        hd: [
          "isFinite - neg_infinity",
          (function (param) {
              return {
                      TAG: /* Eq */0,
                      _0: false,
                      _1: isFinite(Pervasives.neg_infinity)
                    };
            })
        ],
        tl: {
          hd: [
            "isFinite - _NaN",
            (function (param) {
                return {
                        TAG: /* Eq */0,
                        _0: false,
                        _1: isFinite(NaN)
                      };
              })
          ],
          tl: {
            hd: [
              "isFinite - 0.",
              (function (param) {
                  return {
                          TAG: /* Eq */0,
                          _0: true,
                          _1: isFinite(0)
                        };
                })
            ],
            tl: {
              hd: [
                "toExponential",
                (function (param) {
                    return {
                            TAG: /* Eq */0,
                            _0: "1.23456e+2",
                            _1: (123.456).toExponential()
                          };
                  })
              ],
              tl: {
                hd: [
                  "toExponential - large number",
                  (function (param) {
                      return {
                              TAG: /* Eq */0,
                              _0: "1.2e+21",
                              _1: (1.2e21).toExponential()
                            };
                    })
                ],
                tl: {
                  hd: [
                    "toExponentialWithPrecision - digits:2",
                    (function (param) {
                        return {
                                TAG: /* Eq */0,
                                _0: "1.23e+2",
                                _1: (123.456).toExponential(2)
                              };
                      })
                  ],
                  tl: {
                    hd: [
                      "toExponentialWithPrecision - digits:4",
                      (function (param) {
                          return {
                                  TAG: /* Eq */0,
                                  _0: "1.2346e+2",
                                  _1: (123.456).toExponential(4)
                                };
                        })
                    ],
                    tl: {
                      hd: [
                        "toExponentialWithPrecision - digits:20",
                        (function (param) {
                            return {
                                    TAG: /* Eq */0,
                                    _0: "0.00000000000000000000e+0",
                                    _1: (0).toExponential(20)
                                  };
                          })
                      ],
                      tl: {
                        hd: [
                          "File \"js_float_test.ml\", line 31, characters 3-10",
                          (function (param) {
                              return {
                                      TAG: /* ThrowAny */7,
                                      _0: (function (param) {
                                          (0).toExponential(101);
                                          
                                        })
                                    };
                            })
                        ],
                        tl: {
                          hd: [
                            "toExponentialWithPrecision - digits:-1",
                            (function (param) {
                                return {
                                        TAG: /* ThrowAny */7,
                                        _0: (function (param) {
                                            (0).toExponential(-1);
                                            
                                          })
                                      };
                              })
                          ],
                          tl: {
                            hd: [
                              "toFixed",
                              (function (param) {
                                  return {
                                          TAG: /* Eq */0,
                                          _0: "123",
                                          _1: (123.456).toFixed()
                                        };
                                })
                            ],
                            tl: {
                              hd: [
                                "toFixed - large number",
                                (function (param) {
                                    return {
                                            TAG: /* Eq */0,
                                            _0: "1.2e+21",
                                            _1: (1.2e21).toFixed()
                                          };
                                  })
                              ],
                              tl: {
                                hd: [
                                  "toFixedWithPrecision - digits:2",
                                  (function (param) {
                                      return {
                                              TAG: /* Eq */0,
                                              _0: "123.46",
                                              _1: (123.456).toFixed(2)
                                            };
                                    })
                                ],
                                tl: {
                                  hd: [
                                    "toFixedWithPrecision - digits:4",
                                    (function (param) {
                                        return {
                                                TAG: /* Eq */0,
                                                _0: "123.4560",
                                                _1: (123.456).toFixed(4)
                                              };
                                      })
                                  ],
                                  tl: {
                                    hd: [
                                      "toFixedWithPrecision - digits:20",
                                      (function (param) {
                                          return {
                                                  TAG: /* Eq */0,
                                                  _0: "0.00000000000000000000",
                                                  _1: (0).toFixed(20)
                                                };
                                        })
                                    ],
                                    tl: {
                                      hd: [
                                        "toFixedWithPrecision - digits:101",
                                        (function (param) {
                                            return {
                                                    TAG: /* ThrowAny */7,
                                                    _0: (function (param) {
                                                        (0).toFixed(101);
                                                        
                                                      })
                                                  };
                                          })
                                      ],
                                      tl: {
                                        hd: [
                                          "toFixedWithPrecision - digits:-1",
                                          (function (param) {
                                              return {
                                                      TAG: /* ThrowAny */7,
                                                      _0: (function (param) {
                                                          (0).toFixed(-1);
                                                          
                                                        })
                                                    };
                                            })
                                        ],
                                        tl: {
                                          hd: [
                                            "toPrecision",
                                            (function (param) {
                                                return {
                                                        TAG: /* Eq */0,
                                                        _0: "123.456",
                                                        _1: (123.456).toPrecision()
                                                      };
                                              })
                                          ],
                                          tl: {
                                            hd: [
                                              "toPrecision - large number",
                                              (function (param) {
                                                  return {
                                                          TAG: /* Eq */0,
                                                          _0: "1.2e+21",
                                                          _1: (1.2e21).toPrecision()
                                                        };
                                                })
                                            ],
                                            tl: {
                                              hd: [
                                                "toPrecisionWithPrecision - digits:2",
                                                (function (param) {
                                                    return {
                                                            TAG: /* Eq */0,
                                                            _0: "1.2e+2",
                                                            _1: (123.456).toPrecision(2)
                                                          };
                                                  })
                                              ],
                                              tl: {
                                                hd: [
                                                  "toPrecisionWithPrecision - digits:4",
                                                  (function (param) {
                                                      return {
                                                              TAG: /* Eq */0,
                                                              _0: "123.5",
                                                              _1: (123.456).toPrecision(4)
                                                            };
                                                    })
                                                ],
                                                tl: {
                                                  hd: [
                                                    "toPrecisionWithPrecision - digits:20",
                                                    (function (param) {
                                                        return {
                                                                TAG: /* Eq */0,
                                                                _0: "0.0000000000000000000",
                                                                _1: (0).toPrecision(20)
                                                              };
                                                      })
                                                  ],
                                                  tl: {
                                                    hd: [
                                                      "File \"js_float_test.ml\", line 61, characters 3-10",
                                                      (function (param) {
                                                          return {
                                                                  TAG: /* ThrowAny */7,
                                                                  _0: (function (param) {
                                                                      (0).toPrecision(101);
                                                                      
                                                                    })
                                                                };
                                                        })
                                                    ],
                                                    tl: {
                                                      hd: [
                                                        "toPrecisionWithPrecision - digits:-1",
                                                        (function (param) {
                                                            return {
                                                                    TAG: /* ThrowAny */7,
                                                                    _0: (function (param) {
                                                                        (0).toPrecision(-1);
                                                                        
                                                                      })
                                                                  };
                                                          })
                                                      ],
                                                      tl: {
                                                        hd: [
                                                          "toString",
                                                          (function (param) {
                                                              return {
                                                                      TAG: /* Eq */0,
                                                                      _0: "1.23",
                                                                      _1: (1.23).toString()
                                                                    };
                                                            })
                                                        ],
                                                        tl: {
                                                          hd: [
                                                            "toString - large number",
                                                            (function (param) {
                                                                return {
                                                                        TAG: /* Eq */0,
                                                                        _0: "1.2e+21",
                                                                        _1: (1.2e21).toString()
                                                                      };
                                                              })
                                                          ],
                                                          tl: {
                                                            hd: [
                                                              "toStringWithRadix - radix:2",
                                                              (function (param) {
                                                                  return {
                                                                          TAG: /* Eq */0,
                                                                          _0: "1111011.0111010010111100011010100111111011111001110111",
                                                                          _1: (123.456).toString(2)
                                                                        };
                                                                })
                                                            ],
                                                            tl: {
                                                              hd: [
                                                                "toStringWithRadix - radix:16",
                                                                (function (param) {
                                                                    return {
                                                                            TAG: /* Eq */0,
                                                                            _0: "7b.74bc6a7ef9dc",
                                                                            _1: (123.456).toString(16)
                                                                          };
                                                                  })
                                                              ],
                                                              tl: {
                                                                hd: [
                                                                  "toStringWithRadix - radix:36",
                                                                  (function (param) {
                                                                      return {
                                                                              TAG: /* Eq */0,
                                                                              _0: "3f",
                                                                              _1: (123).toString(36)
                                                                            };
                                                                    })
                                                                ],
                                                                tl: {
                                                                  hd: [
                                                                    "toStringWithRadix - radix:37",
                                                                    (function (param) {
                                                                        return {
                                                                                TAG: /* ThrowAny */7,
                                                                                _0: (function (param) {
                                                                                    (0).toString(37);
                                                                                    
                                                                                  })
                                                                              };
                                                                      })
                                                                  ],
                                                                  tl: {
                                                                    hd: [
                                                                      "toStringWithRadix - radix:1",
                                                                      (function (param) {
                                                                          return {
                                                                                  TAG: /* ThrowAny */7,
                                                                                  _0: (function (param) {
                                                                                      (0).toString(1);
                                                                                      
                                                                                    })
                                                                                };
                                                                        })
                                                                    ],
                                                                    tl: {
                                                                      hd: [
                                                                        "toStringWithRadix - radix:-1",
                                                                        (function (param) {
                                                                            return {
                                                                                    TAG: /* ThrowAny */7,
                                                                                    _0: (function (param) {
                                                                                        (0).toString(-1);
                                                                                        
                                                                                      })
                                                                                  };
                                                                          })
                                                                      ],
                                                                      tl: {
                                                                        hd: [
                                                                          "fromString - 123",
                                                                          (function (param) {
                                                                              return {
                                                                                      TAG: /* Eq */0,
                                                                                      _0: 123,
                                                                                      _1: Number("123")
                                                                                    };
                                                                            })
                                                                        ],
                                                                        tl: {
                                                                          hd: [
                                                                            "fromString - 12.3",
                                                                            (function (param) {
                                                                                return {
                                                                                        TAG: /* Eq */0,
                                                                                        _0: 12.3,
                                                                                        _1: Number("12.3")
                                                                                      };
                                                                              })
                                                                          ],
                                                                          tl: {
                                                                            hd: [
                                                                              "fromString - empty string",
                                                                              (function (param) {
                                                                                  return {
                                                                                          TAG: /* Eq */0,
                                                                                          _0: 0,
                                                                                          _1: Number("")
                                                                                        };
                                                                                })
                                                                            ],
                                                                            tl: {
                                                                              hd: [
                                                                                "fromString - 0x11",
                                                                                (function (param) {
                                                                                    return {
                                                                                            TAG: /* Eq */0,
                                                                                            _0: 17,
                                                                                            _1: Number("0x11")
                                                                                          };
                                                                                  })
                                                                              ],
                                                                              tl: {
                                                                                hd: [
                                                                                  "fromString - 0b11",
                                                                                  (function (param) {
                                                                                      return {
                                                                                              TAG: /* Eq */0,
                                                                                              _0: 3,
                                                                                              _1: Number("0b11")
                                                                                            };
                                                                                    })
                                                                                ],
                                                                                tl: {
                                                                                  hd: [
                                                                                    "fromString - 0o11",
                                                                                    (function (param) {
                                                                                        return {
                                                                                                TAG: /* Eq */0,
                                                                                                _0: 9,
                                                                                                _1: Number("0o11")
                                                                                              };
                                                                                      })
                                                                                  ],
                                                                                  tl: {
                                                                                    hd: [
                                                                                      "fromString - invalid string",
                                                                                      (function (param) {
                                                                                          return {
                                                                                                  TAG: /* Eq */0,
                                                                                                  _0: true,
                                                                                                  _1: isNaN(Number("foo"))
                                                                                                };
                                                                                        })
                                                                                    ],
                                                                                    tl: /* [] */0
                                                                                  }
                                                                                }
                                                                              }
                                                                            }
                                                                          }
                                                                        }
                                                                      }
                                                                    }
                                                                  }
                                                                }
                                                              }
                                                            }
                                                          }
                                                        }
                                                      }
                                                    }
                                                  }
                                                }
                                              }
                                            }
                                          }
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
};

var suites = {
  hd: suites_0,
  tl: suites_1
};

Mt.from_pair_suites("Js_float_test", suites);

exports.suites = suites;
/*  Not a pure module */
